src/pstl_offload.cpp

```diff
@@ -96,59 +97,311 @@ __original_msize(void* __user_ptr)
+        __orig_free_allocator<std::pair<void* const, __ptr_desc>>>;
+    __map_ptr_to_object_prop* _M_map;
+
+public:
+    // We suppose that all users of libpstloffload have dependence on it, so it's impossible to
+    // register >=4K-aligned USM memory before ctor of static objects in libpstloffload is executed.
+    // So, no need for special support for adding to not-yet-created __large_aligned_ptrs_map.
+    __large_aligned_ptrs_map()
+    {
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        _M_map = new __map_ptr_to_object_prop;
+    }
+
+    // Do not destroy (i.e., intentionally leak) _M_map to able use it after static object dtor is
+    // executed. Global free/delete/realloc/etc are overloaded, so we need to use it even after
+    // static object dtor has been executed.
+    ~__large_aligned_ptrs_map() { }
+
+    void
+    __register_ptr(void* __ptr, std::size_t __size, __sycl_device_shared_ptr __device_ptr)
+    {
+        assert(__is_ptr_page_aligned(__ptr));
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        [[maybe_unused]] auto __ret = _M_map->emplace(__ptr, __ptr_desc{std::move(__device_ptr), __size});
+        assert(__ret.second); // the pointer must be unique
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<__ptr_desc>
+    __unregister_ptr(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered, so they may not be in the map
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        __ptr_desc __header = std::move(__iter->second);
+        __this._M_map->erase(__iter);
+        return __header;
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<std::size_t>
+    __get_size(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        return __iter->second._M_requested_number_of_bytes;
+    }
+};
+
+static __large_aligned_ptrs_map __large_aligned_ptrs;
+
+void
+__block_header::__free()
```

Comment:
Why did you decide to make it a member function?

Reply:
Let's make it just function, if you prefer.

Reply:
ok, a different and more important question. Why did we move it to the cpp part? is it because it is not used in headers anymore?